velox/functions/lib/Re2Functions.cpp

```diff
@@ -1031,28 +1122,42 @@ PatternMetadata determinePatternKind(StringView pattern) {
   // At this point pattern contains exactly one fixed pattern.
   // Pattern contains no wildcard characters (is a fixed pattern).
   if (wildcardStart == -1) {
-    return PatternMetadata{PatternKind::kFixed, patternLength};
+    auto unescapedPattern = unescape(pattern, 0, patternLength, escapeChar);
+    return PatternMetadata{
+        PatternKind::kFixed,
+        (vector_size_t)unescapedPattern.size(),
+        unescapedPattern};
   }
+
   // Pattern is generic if it has '_' wildcard characters and a fixed pattern.
   if (singleCharacterWildcardCount) {
     return PatternMetadata{PatternKind::kGeneric, 0};
   }
   // Classify pattern as prefix, fixed center, or suffix pattern based on the
   // position and count of the wildcard character sequence and fixed pattern.
   if (fixedPatternStart < wildcardStart) {
-    return PatternMetadata{PatternKind::kPrefix, wildcardStart};
+    auto unescapedPattern = unescape(pattern, 0, wildcardStart, escapeChar);
+    return PatternMetadata{
+        PatternKind::kPrefix,
+        (vector_size_t)unescapedPattern.size(),
+        unescapedPattern};
   }
   // if numWildcardSequences > 1, then fixed pattern must be in between them.
   if (numWildcardSequences == 2) {
+    auto unescapedPattern =
+        unescape(pattern, fixedPatternStart, fixedPatternEnd + 1, escapeChar);
     return PatternMetadata{
         PatternKind::kSubstring,
-        0,
-        std::string(
-            pattern.data() + fixedPatternStart,
-            fixedPatternEnd + 1 - fixedPatternStart)};
+        (vector_size_t)unescapedPattern.size(),
+        unescapedPattern};
   }
+  auto unescapedPattern =
+      unescape(pattern, fixedPatternStart, patternLength, escapeChar);
+
   return PatternMetadata{
-      PatternKind::kSuffix, patternLength - fixedPatternStart};
+      PatternKind::kSuffix,
+      (vector_size_t)unescapedPattern.size(),
```

Comment:
Why do we need to pass both unescapedPattern and its size?

Reply:
Previously, when we do not support escape char, we don't need to pass `pattern` to `PatternMetadata`, because we can deduce the `suffix pattern` from the original `pattern` and `size`, take `%ab` as an example:

```
pattern = %ab, size = 2   -> suffix_pattern = ab
```

now we support escape char, with only size, we can not deduce the actual `suffix_pattern`, take `%a\_b` as example(`\` is the escape char), since the unescaped pattern is `a_b`, the size is `3`, but if we get the last 3 letters from the *original pattern*, it is `\_b`, which is wrong, so we need to pass both unescaped pattern and size.